akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala

```diff
@@ -61,21 +61,17 @@ private[projection] abstract class InternalProjectionState[Offset, Envelope](
   val killSwitch: SharedKillSwitch = KillSwitches.shared(projectionId.id)
   val abort: Promise[Done] = Promise()
 
-  private def reportProgress[T](after: Future[T], envelope: Envelope): Future[T] = {
-    after
+  private def saveOffsetAndReport(
+      projectionId: ProjectionId,
+      projectionContext: ProjectionContextImpl[Offset, Envelope],
+      batchSize: Int): Future[Done] = {
+    saveOffset(projectionId, projectionContext.offset)
       .map { done =>
         try {
-          statusObserver.progress(projectionId, envelope)
+          statusObserver.offsetProgress(projectionId, projectionContext.envelope)
         } catch {
           case NonFatal(_) => // ignore
         }
-        done
-      }
-  }
-
-  private def saveOffsetWithMetrics(projectionId: ProjectionId, offset: Offset, batchSize: Int): Future[Done] = {
-    saveOffset(projectionId, offset)
-      .map { done =>
         telemetry.onOffsetStored(batchSize)
         done
       }
```

Comment:
the change in semantics made `reportProgress` irrelevant. Instead, `statusObserver.offsetProgress` is invoked immediately before `telemetry.onOffsetStored` in all cases.